File: solutions/go/sublist/1/sublist.go
package sublist

import "reflect"

type Relation string

func Sublist(one, two []int) Relation {
	switch {
	case equal(one, two):
		return "equal"
	case sublist(one, two):
		return "sublist"
	case superlist(one, two):
		return "superlist"
	}
	return "unequal"
}

func equal(one, two []int) bool {
	if len(one) != len(two) {
		return false
	}
	return reflect.DeepEqual(one, two)
}

func sublist(one, two []int) bool {
	if len(one) == 0 {
		return true
	}
	index := -1
	for i := range two {
		if two[i] == one[0] {
			index = i
			break
		}
	}
	if index == -1 {
		return false
	}
	one_end := index + len(one)
	if one_end > len(two) {
		return false
	}
	if reflect.DeepEqual(one, two[index:one_end]) {
		return true
	}
	return sublist(one, two[index+1:])
}

func superlist(one, two []int) bool {
	return sublist(two, one)
}

File: solutions/zig/sublist/1/sublist.zig
const std = @import("std");

pub const Relation = enum { equal, sublist, superlist, unequal };

const T = i32;

pub fn compare(list_one: []const T, list_two: []const T) Relation {
    if (std.mem.eql(T, list_one, list_two)) return .equal;
    if (contains(list_one, list_two)) return .superlist;
    if (contains(list_two, list_one)) return .sublist;
    return .unequal;
}

fn contains(list_one: []const T, list_two: []const T) bool {
    if (list_one.len == 0) return false;

    return std.mem.startsWith(T, list_one, list_two) or contains(list_one[1..], list_two);
}
